File: iOS/TreasureHunt/TreasureHuntObjc/HuntViewController.m
//
//  HuntViewController.m
//  TreasureHuntObjc
//
//  Created by Seamus on 05/11/2016.
//
//

#import "HuntViewController.h"

@interface HuntViewController () {
    int numberOfLocationRequests;
}

@end

@implementation HuntViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self.locationManager setDelegate:self];
    numberOfLocationRequests = 0;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}

 
 */

-(IBAction)tryLocationButton:(id)sender {
    UIAlertController* alert = [UIAlertController alertControllerWithTitle:@"Fee" message:@"" preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *defaultAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        NSLog(@"API CALL");
        [self.locationManager requestLocation];
        numberOfLocationRequests = 1;
    }];
    UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {
        
    }];
    [alert addAction:defaultAction];
    [alert addAction:cancelAction];
    [self presentViewController:alert animated:YES completion:nil];
}

-(void) didUpdateLocation:(CLLocation *)location {
    if(numberOfLocationRequests > 0) {
         NSURLSession* ourSession = [NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]];
        
        NSMutableURLRequest* someRequest = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:@"http://46.101.80.224/"]];
        [someRequest setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
        [someRequest setValue:@"application/json" forHTTPHeaderField:@"Accept"];
        [someRequest setHTTPMethod:@"POST"];
        
        
        [someRequest setHTTPBody:[NSJSONSerialization dataWithJSONObject:@{@"latitude":[NSNumber numberWithDouble:location.coordinate.latitude], @"longitude":[NSNumber numberWithDouble: location.coordinate.longitude]} options:0 error:nil] ];
        
        
        NSURLSessionTask *task = [ourSession dataTaskWithRequest:someRequest completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
            if (response != nil ) {
                NSLog(@"%@", response);
            } else {
                NSLog(@"%@", error.localizedDescription);
            }
        }];
        [task resume];
    }
    numberOfLocationRequests--;
}

-(void) didFail {
    
}


@end

File: iOS/TreasureHunt/TreasureHunt/HuntViewController.swift
//
//  HuntViewController.swift
//  TreasureHunt
//
//  Created by Tim Colla on 05/11/2016.
//  Copyright © 2016 Marinosoftware. All rights reserved.
//

import UIKit
import CoreLocation

class HuntViewController: UIViewController, LocationManagerDelegate {
    @IBOutlet weak var hintLabel: UILabel!
    
    let locationManager = LocationManager()
    
    override func viewDidLoad() {
        locationManager.delegate = self
    }
    
    @IBAction func tryLocationButton(_ sender: UIButton) {
        /*UIAlertController* alert = [UIAlertController alertControllerWithTitle:@"My Alert"
        message:@"This is an alert."
        preferredStyle:UIAlertControllerStyleAlert];
        
        UIAlertAction* defaultAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault
        handler:^(UIAlertAction * action) {}];
        
        [alert addAction:defaultAction];
        [self presentViewController:alert animated:YES completion:nil];*/
        
        let alert = UIAlertController(title: "Fee", message: "Trying a location will cost a 40000 Satoshi fee.", preferredStyle: .alert)
        let defaultAction = UIAlertAction(title: "OK", style: .default) { (action) in
            print("API call")
            
            self.locationManager.requestLocation()
        }
        let cancelAction = UIAlertAction(title: "Cancel", style: .default) { (action) in
            alert.dismiss(animated: true, completion: {})
        }
        alert.addAction(defaultAction)
        alert.addAction(cancelAction)
        
        self.present(alert, animated: true) {}
    }
    
    func didUpdate(location: CLLocation) {
        let ourSession = URLSession(configuration: .default)
        var someRequest = URLRequest(url: URL(string:"http://46.101.80.224/")!)
        
        someRequest.addValue("application/json", forHTTPHeaderField: "Content-Type")
        someRequest.addValue("application/json", forHTTPHeaderField: "Accept")
        someRequest.httpMethod = "POST"
        someRequest.httpBody = try! JSONSerialization.data(withJSONObject: ["latitude":location.coordinate.latitude, "longitude":location.coordinate.longitude], options: [])
        
        let task = ourSession.dataTask(with: someRequest) {
            (data, response, error) in
            
            if let response = response as? HTTPURLResponse {
                print(response)
            } else {
                
                if let error = error {
                    print(error)
                }
            }
        }
        
        task.resume()
    }
    
    func didFail() {
        
    }
}
